streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/bolt/query/RealtimeJoinBolt.java

```diff
@@ -0,0 +1,638 @@
+    protected Object findField(FieldSelector fieldSelector, Tuple tuple) {
+        if (tuple==null) {
+            return null;
+        }
+        // very stream name matches, it stream name was specified
+        if ( fieldSelector.streamName!=null &&
+                !fieldSelector.streamName.equalsIgnoreCase( getStreamSelector(tuple) ) ) {
+            return null;
+        }
+
+        Object curr = null;
+        for (int i=0; i < fieldSelector.field.length; i++) {
+            if (i==0) {
+                if (tuple.contains(fieldSelector.field[i]) )
+                    curr = tuple.getValueByField(fieldSelector.field[i]);
+                else
+                    return null;
+            }  else  {
+                curr = ((Map) curr).get(fieldSelector.field[i]);
+                if (curr==null)
+                    return null;
+            }
+        }
+        return curr;
+    }
+
+    /** Performs projection on the tuples based on 'projectionFields'
+     *
+     * @param tuple1   can be null
+     * @param tuple2   can be null
+     * @return   project fields
+     */
+
+    protected ArrayList<Object> doProjection(Tuple tuple1, Tuple tuple2) {
+        if(streamLineProjection)
+            return doStreamlineProjection(tuple1, tuple2);
+
+        ArrayList<Object> result = new ArrayList<>(outputFields.length);
+        for ( int i = 0; i < outputFields.length; i++ ) {
+            FieldSelector outField = outputFields[i];
+            Object field = findField(outField, tuple1) ;
+            if (field==null)
+                field = findField(outField, tuple2);
+            result.add(field); // adds null if field is not found in both tuples
+        }
+        return result;
+    }
+
+    // NOTE: Streamline specific convenience method. Creates output tuple as a StreamlineEvent
+    protected ArrayList<Object> doStreamlineProjection(Tuple tuple1, Tuple tuple2) {
+//        String flattenedKey = projectionKeys[i].getOutputName();
+//        String outputKeyName = dropStreamLineEventPrefix(flattenedKey); // drops the "streamline-event." prefix
+        StreamlineEventImpl.Builder eventBuilder = StreamlineEventImpl.builder();
+
+        ArrayList<Object> result = new ArrayList<>(outputFields.length);
+        for ( int i = 0; i < outputFields.length; i++ ) {
+            FieldSelector outField = outputFields[i];
+
+            Object field = findField(outField, tuple1) ;
+            if (field==null)
+                field = findField(outField, tuple2);
+            String outputKeyName = dropStreamLineEventPrefix(outField.getOutputName() );
+            eventBuilder.put(outputKeyName, field); // adds null if field is not found in both tuples
+        }
+
+        ArrayList<Object> resultRow = new ArrayList<>();
+        StreamlineEventImpl slEvent = eventBuilder.dataSourceId("multiple sources").build();
+        resultRow.add(slEvent);
+        return resultRow;
```

Comment:
may want to return singleton list like below instead of creating a new ArrayList and adding one entry.
```
Collections.singletonList(eventBuilder.dataSourceId("multiple sources").build())
```